cdap-app-fabric/src/main/java/co/cask/cdap/scheduler/ScheduleNotificationSubscriberService.java

```diff
@@ -179,6 +196,57 @@ public void processNotification(DatasetContext context, Notification notificatio
     }
   }
 
+  /**
+   * Private class that receives program status notifications forwarded from
+   * {@link co.cask.cdap.internal.app.services.ProgramNotificationSubscriberService}.
+   *
+   * Processes notifications that are guaranteed to trigger a program
+   */
+  private class ProgramStatusEventNotificationSubscriberThread extends SchedulerEventNotificationSubscriberThread {
+    ProgramStatusEventNotificationSubscriberThread() {
+      super(cConf.get(Constants.AppFabric.PROGRAM_STATUS_EVENT_TOPIC));
+    }
+
+    @Override
+    public void processNotification(DatasetContext context, Notification notification)
+      throws IOException, DatasetManagementException {
+
+      String programRunIdString = notification.getProperties().get(ProgramOptionConstants.PROGRAM_RUN_ID);
+      String programRunStatusString = notification.getProperties().get(ProgramOptionConstants.PROGRAM_STATUS);
+
+      ProgramStatus programStatus;
+      try {
+        programStatus = ProgramRunStatus.toProgramStatus(ProgramRunStatus.valueOf(programRunStatusString));
+      } catch (IllegalArgumentException e) {
+        // Return silently, this happens for statuses that are not meant to be scheduled
+        return;
+      }
+
+      // Ignore notifications which specify an invalid programRunId or programStatus
+      if (programRunIdString == null || programStatus == null) {
+        return;
+      }
+
+      ProgramRunId programRunId = GSON.fromJson(programRunIdString, ProgramRunId.class);
+      ProgramId programId = programRunId.getParent();
+      String triggerKeyForProgramStatus = Schedulers.triggerKeyForProgramStatus(programId, programStatus);
+      Collection<ProgramScheduleRecord> triggeredSchedules = getSchedules(context, triggerKeyForProgramStatus);
+
+      // Only fetch the triggering program information if there is at least one triggered and enabled schedule
```

Comment:
I see, I think we should check it in the `addNotification` method though. `SchedulerEventNotificationSubscriberThread` currently misses this check